nemo_curator/stages/text/embedders/vllm.py

```diff
@@ -0,0 +1,126 @@
+        text_field: The field in the DataFrame containing the text to embed.
+        embedding_field: The field name for the output embeddings.
+        hf_token: Hugging Face token for downloading the model, if needed.
+        gpu_memory_utilization: Fraction of GPU memory to use for VLLM.
+        num_cpus: Number of CPUs to use for the model.
+        num_gpus: Number of GPUs to use for the model.
+
+    """
+
+    def __init__(  # noqa: PLR0913
+        self,
+        model_identifier: str,
+        text_field: str = "text",
+        embedding_field: str = "embeddings",
+        hf_token: str | None = None,
+        gpu_memory_utilization: float = 0.5,
+        num_cpus: int = 8,
+        num_gpus: int = 1,
+    ):
+        # Initialize base ModelStage - VLLM doesn't use tokenized input or batching
+        super().__init__(
+            model_identifier=model_identifier,
+            hf_token=hf_token,
+            has_seq_order=False,  # VLLM preserves order internally
+        )
+        if num_gpus > 1:
+            msg = "Tensor parallelism is not supported for VLLM yet. If you need it, please open an issue on GitHub."
+            raise ValueError(msg)
+
+        # Override name and resources for VLLM-specific configuration
+        self.name = format_name_with_suffix(model_identifier, suffix="_vllm_model")
+        self.resources = Resources(cpus=num_cpus, gpus=num_gpus)
+
+        self.text_field = text_field
+        self.embedding_field = embedding_field
+        self.gpu_memory_utilization = gpu_memory_utilization
+        self.num_gpus = num_gpus
+        self.num_cpus = num_cpus
+        self.model = None
+
+    def inputs(self) -> tuple[list[str], list[str]]:
+        return ["data"], [self.text_field]
+
+    def outputs(self) -> tuple[list[str], list[str]]:
+        return ["data"], [self.embedding_field]
+
+    def _setup(self, local_files_only: bool = True) -> None:
+        """Load the VLLM model for inference."""
+
+        if self.model is None:
+            start_time = time.time()
+            self.model = vllm.LLM(
+                model=self.model_identifier,
+                runner="pooling",
+                gpu_memory_utilization=self.gpu_memory_utilization,
+                tensor_parallel_size=max(1, int(self.num_gpus)),
+                trust_remote_code=not local_files_only,
+            )
+
+            end_time = time.time()
+            logger.info(f"VLLM model loaded in {end_time - start_time} seconds")
+        else:
+            logger.info("VLLM model already loaded")
+
+    def process(self, batch: DocumentBatch) -> DocumentBatch:
+        """Process all texts in a single VLLM call.
+
+        VLLM handles batching internally, so we submit all texts at once
+        for optimal throughput.
+        """
+        df_cpu = batch.to_pandas()
+
+        # Get all texts and submit to VLLM at once
+        texts = df_cpu[self.text_field].tolist()
```

Comment:
Consider supporting `max_chars`:
```suggestion
        texts = df_cpu[self.text_field].tolist()

        if self.max_chars is not None:
            texts = [texts[:self.max_chars] for text in texts]
```